<a href="https://colab.research.google.com/github/ramirezcordero7lp/lenguajesdeprogramacion/blob/main/Trabajo_1_Lenguaje_Programaci%C3%B3n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Caso práctico: Analista de inversiones

Imagina que trabajas como **analista financiero** en una empresa de inversión.

Tu equipo debe decidir en cuál proyecto invertir.

La empresa solo puede elegir **un proyecto**.

Tu tarea será usar **Python para analizar la rentabilidad de cada proyecto** utilizando el **Valor Presente Neto (VPN)**.

## ¿Qué es el Valor Presente Neto (VPN)?

El **Valor Presente Neto (VPN)** permite determinar si un proyecto de inversión es rentable.

- Si **VPN > 0** → el proyecto genera valor.
- Si **VPN < 0** → el proyecto no es rentable.

Fórmula:

$$VPN = \\sum \
rac{Flujo_t}{(1+r)^t} - Inversión\ Inicial$$

## Proyectos de inversión

Usaremos **diccionarios en Python** para representar cada proyecto.

In [ ]:
proyecto_Carlos= {
    'nombre': 'Proyecto Carlos Ltda',
    'inversion': 8000,
    'flujos': [2500, 3000, 3500, 4000],
    'tasa': 0.10
}

proyecto_Luisa = {
    'nombre': 'Proyecto Luisa SAS',
    'inversion': 10000,
    'flujos': [3000, 3500, 4200, 4500],
    'tasa': 0.10
}

proyecto_Liz = {
    'nombre': 'Proyecto Liz SAS',
    'inversion': 7000,
    'flujos': [1800, 2600, 3200, 3800],
    'tasa': 0.10
}

## Calcular el VPN usando `for`

## Ejercicio

Calcule el VPN del **Proyecto B** y **Proyecto C**.

In [ ]:
vpn = -proyecto_Carlos['inversion']

for t, flujo in enumerate(proyecto_Carlos['flujos']):
    vpn += flujo / (1 + proyecto_Carlos['tasa'])**(t+1)

print('VPN del proyecto A:', round(vpn, 3))

VPN del proyecto A: 2113.722


In [ ]:
if vpn > 25680:
  print('Muy bien tu proyecto es rentable felicitaciones.')
else:
    print('Ojo que tu proyecto no es rentable, por lo tanto te invito a buscar otra alternativa')

Ojo que tu proyecto no es rentable, por lo tanto te invito a buscar otra alternativa


In [ ]:
vpn = -proyecto_Luisa['inversion']

for t, flujo in enumerate(proyecto_Luisa['flujos']):
  vpn += flujo / (1 + proyecto_Luisa['tasa'])**(t+1)

  print('VPN del proyecto Luisa:', round(vpn, 3))


VPN del proyecto Luisa: -7272.727
VPN del proyecto Luisa: -4380.165
VPN del proyecto Luisa: -1224.643
VPN del proyecto Luisa: 1848.917


In [ ]:
if vpn > 34587:
    print('Genial!!; tú proyecto es rentable')
else:
    print('Upss, tu proyecto no es rentable')

Upss, tu proyecto no es rentable


In [ ]:
vpn = -proyecto_Liz['inversion']

for t, flujo in enumerate(proyecto_Liz['flujos']):
  vpn += flujo / (1 + proyecto_Liz['tasa'])**(t+1)

  print('VPN del proyecto Liz:', round(vpn, 3))

VPN del proyecto Liz: -5363.636
VPN del proyecto Liz: -3214.876
VPN del proyecto Liz: -810.669
VPN del proyecto Liz: 1784.782


In [ ]:
if vpn > 1000:
    print('Eyyyy, tú proyecto es rentable')
else:
    print('Uhhh, tu proyecto no es rentable')

Eyyyy, tú proyecto es rentable


## Decisión con `if`

## Analizar todos los proyectos

In [ ]:
proyectos = [proyecto_Carlos, proyecto_Luisa, proyecto_Liz]

for proyecto in proyectos:
    vpn = -proyecto['inversion']

    for t, flujo in enumerate(proyecto['flujos']):
        vpn += flujo / (1 + proyecto['tasa'])**(t+1)

    print(proyecto['nombre'], 'VPN:', round(vpn, 2))

Proyecto Carlos Ltda VPN: 2113.72
Proyecto Luisa SAS VPN: 1848.92
Proyecto Liz SAS VPN: 1784.78


## Simulación de tasas con `while`

In [ ]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_Carlos['inversion']

    for t, flujo in enumerate(proyecto_Carlos['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 3416.28
Tasa: 0.1 VPN: 2113.72
Tasa: 0.15000000000000002 VPN: 1030.66
Tasa: 0.2 VPN: 121.14


In [ ]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_Luisa['inversion']

    for t, flujo in enumerate(proyecto_Luisa['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 3362.03
Tasa: 0.1 VPN: 1848.92
Tasa: 0.15000000000000002 VPN: 589.66
Tasa: 0.2 VPN: -468.75


In [ ]:
tasa = 0.05

while tasa <= 0.20:
    vpn = -proyecto_Liz['inversion']

    for t, flujo in enumerate(proyecto_Liz['flujos']):
        vpn += flujo / (1 + tasa)**(t+1)

    print('Tasa:', tasa, 'VPN:', round(vpn, 2))

    tasa += 0.05

Tasa: 0.05 VPN: 2963.11
Tasa: 0.1 VPN: 1784.78
Tasa: 0.15000000000000002 VPN: 807.91
Tasa: 0.2 VPN: -10.03


## Reto final

1. Calcule el VPN de los tres proyectos.
2. Determine cuál proyecto tiene el **VPN más alto**.
3. Justifique la decisión.

2. Determinar cuál proyecto tiene el VPN mas alto

El VPN más alto lo tiene el proyecto Carlos con un valor de 2113.72 en comparación con los otros dos proyectos que tienen un VPN inferior a 2000, pero los 3 proyectos tiene VPN positiva lo que los hace viables.

3. Justifique la decisión

El VPN del proyecto Carlos al ser mas alto generara mayor valor economico a la empresa, este se vuelve la mejor alternativa de inversión porque proporciona mayor rentabilidad y mayor incremento en el valor del dinero invertido.